auth/crammd5/authenticatee.go

```diff
@@ -0,0 +1,236 @@
+package crammd5
+
+import (
+	"fmt"
+	"sync"
+
+	"code.google.com/p/gogoprotobuf/proto"
+	log "github.com/golang/glog"
+	"github.com/mesos/mesos-go/auth"
+	mesos "github.com/mesos/mesos-go/mesosproto"
+	"github.com/mesos/mesos-go/messenger"
+	"github.com/mesos/mesos-go/upid"
+	"golang.org/x/net/context"
+)
+
+var (
+	pidLock        sync.Mutex
+	pid            int
+	supportedMechs map[string]struct{}
+)
+
+type statusType int
+
+const (
+	READY statusType = iota
+	STARTING
+	STEPPING
+	COMPLETED
+	FAILED
+	ERROR
+	DISCARDED
+)
+
+type authenticateeProcess struct {
+	*messenger.MesosMessenger
+	credential mesos.Credential
+	client     upid.UPID
+	status     statusType
+	initOnce   sync.Once
+	done       chan struct{}
+	result     error
+	ctx        context.Context
+	cancel     context.CancelFunc
+	mech       *mechanism
+	stepFn     stepFunc
+}
+
+func init() {
+	supportedMechs = make(map[string]struct{})
+	supportedMechs["CRAM-MD5"] = struct{}{}
+}
+
+func nextPid() int {
+	pidLock.Lock()
+	defer pidLock.Unlock()
+	pid++
+	return pid
+}
+
+func Authenticatee(pid, client upid.UPID, credential mesos.Credential) <-chan error {
```

Comment:
It's not a constructor function - just a top-level func in the package to handle the client side of the SASL authentication process. For comparison `newAuthenticatee` is a constructor function.